Library/Homebrew/cmd/bottle.rb

```diff
@@ -300,74 +301,104 @@ def bottle_formula(f)
-
-    merge_hash = {}
-    ARGV.named.each do |argument|
-      bottle_block = IO.read(argument)
-      formula_name = bottle_block.lines.first.sub(/^# /, "").chomp
-      merge_hash[formula_name] ||= []
-      merge_hash[formula_name] << bottle_block
+
+    bottles_hash = ARGV.named.reduce({}) do |hash, json_file|
+      hash.merge! Utils::JSON.load(IO.read(json_file))
     end
 
-    merge_hash.each do |formula_name, bottle_blocks|
+    bottles_hash.each do |formula_name, bottle_hash|
       ohai formula_name
-      f = Formulary.factory(formula_name)
 
-      if f.bottle_disabled?
-        ofail "Formula #{f.full_name} has disabled bottle"
-        puts f.bottle_disable_reason
-        next
+      bottle = BottleSpecification.new
+      bottle.root_url bottle_hash["bottle"]["root_url"]
+      cellar = bottle_hash["bottle"]["cellar"]
+      if cellar == "any" || cellar == "any_skip_relocation"
+        cellar = cellar.to_sym
       end
-
-      bottle = if keep_old
-        f.bottle_specification.dup
-      else
-        BottleSpecification.new
-      end
-      bottle.extend(BottleMerger)
-      bottle_blocks.each { |block| bottle.instance_eval(block) }
-
-      old_spec = f.bottle_specification
-      if keep_old && !old_spec.checksums.empty?
-         bad_fields = [:root_url, :prefix, :cellar, :revision].select do |field|
-           old_spec.send(field) != bottle.send(field)
-         end
-         bad_fields.delete(:cellar) if old_spec.cellar == :any && bottle.cellar == :any_skip_relocation
-         if bad_fields.any?
-           ofail "--keep-old is passed but there are changes in: #{bad_fields.join ", "}"
-           next
-         end
+      bottle.cellar cellar
+      bottle.prefix bottle_hash["bottle"]["prefix"]
+      bottle.revision bottle_hash["bottle"]["revision"]
+      bottle_hash["bottle"]["tags"].each do |tag, tag_hash|
+        bottle.sha256 tag_hash["sha256"] => tag.to_sym
       end
 
       output = bottle_output bottle
-      puts output
 
       if write
+        path = Pathname.new("#{HOMEBREW_REPOSITORY/bottle_hash["formula"]["path"]}")
         update_or_add = nil
 
-        Utils::Inreplace.inreplace(f.path) do |s|
+        Utils::Inreplace.inreplace(path) do |s|
           if s.include? "bottle do"
             update_or_add = "update"
+            if ARGV.include? "--keep-old"
+              mismatches = []
+              s =~ /  bottle do(.+?)end\n/m
+              bottle_block_contents = $1
+              bottle_block_contents.lines.each do |line|
+                line = line.strip
+                next if line.empty?
+                key, value, _, tag = line.split " ", 4
+                value = value.to_s.delete ":'\""
+                tag = tag.to_s.delete ":"
+
+                if !tag.empty?
+                  if !bottle_hash["bottle"]["tags"][tag].to_s.empty?
+                    mismatches << "#{key} => #{tag}"
+                  else
+                    bottle.send(key, value => tag.to_sym)
```

Comment:
I think both of them can be bypassed.
* By insert code in formula file, I think it's quite possible to make `test-bot` think any PR as a `--keep-old` situation, and trigger `keep-old` for downstream bottle upload job.
* By make `key` as `eval` or other similar dangerous method, it will be able to run arbitrary code.